design-proposals/instancetype.kubevirt.io/pkg-instancetype-refactor.md

```diff
@@ -0,0 +1,160 @@
+
+## Move store code under `pkg/{instancetype,preference}/store`
+
+Any common code placed under `pkg/instancetype`.
+
+## Extract webhook validation code under `pkg/{instancetype,preference}/validate`
+
+Moving this code under `pkg/{instancetype,preference}` will ease future maintenance.
+
+## Replace use of `InstancetypeMethods` with specific interfaces per consumer requirements
+
+At present the `InstancetypeMethods` interface is requested by various webhooks, subresource APIs and controllers:
+
+* `VMsAdmitter`  
+* `VMsMutator`  
+* `SubresourceAPIApp`  
+* `VMController`  
+* `VMExportController`
+
+There are also several other locations where public functions from `pkg/instancetype` are called directly.
+
+The use of a single public interface exported by `pkg/instancetype` was a mistake.
+
+This should be replaced with smaller simple private interface definitions alongside consumers of the implementations provided by `pkg/{instancetype,preference}`.
+
+Once replaced, `InstancetypeMethods` can be removed from `pkg/instancetype` entirely.
+
+## API Examples
+
+Taking a very simplistic example from  `VMController`, we can replace the use of `InstancetypeMethods` with a private interface defined to meet the requirements of the specific consumer.
+
+For the sake of this example the only requirement for the handler is to find an instance type, thus we can define the following interface within the controller:
+
+```go
+type instancetypeHandler interface {
+    Find(vm *v1.VirtualMachine) (*v1beta1.VirtualMachineInstancetypeSpec, error)
+}
+```
+
+As set out above the implementation of this logic will reside under `pkg/instancetype/find` and will satisfy this interface.
+
+```go
+type Finder struct {
+    Store                   cache.Store
+    ClusterStore            cache.Store
+    ControllerRevisionStore cache.Store
+    Client                  kubevirtcli.KubevirtClient
+}
+
+func (f *Finder) Find(vm *v1.VirtualMachine) (*v1beta1.VirtualMachineInstancetypeSpec, error) {
+ [..]
+}
+```
+
+We can then pull in the required implementation into a handler in order to satisfy the interface within the controller:
+
+```go
+import "kubevirt.io/kubevirt/pkg/instancetype/find"
+
+type instancetypeHandler interface {
+    Find(vm *v1.VirtualMachine) (*v1beta1.VirtualMachineInstancetypeSpec, error)
+}
+
+type handler struct {
+    find.Finder
+}
+
+var _ instancetypeHandler := &handler{}
+
+instancetypeHandler := &handler{
+    Finder: instancetypeFind.Finder{
+        Store:                  instancetypeStore,
+        ClusterStore:           clusterInstancetypeStore,
+        ControllerRevisionStore controllerRevisionStore,
+        Client:                 virtClient,
+    },
+}
+
+c := &VMController{
+    [..]
+    instancetypeHandler: instancetypeHandler
+    [..]
+}
+```
+
+Before using this handler throughout the controller:
+
+```go
+func (c *VMController) Sync(){
+    [..]
+    spec, err := c.instancetypeHandler.Find(vm)
+    [..]
+}
+
+```
+
```

Comment:
Bonus: No more stubs required in tests.

Reply:
I don't see how this changes anything tbh, we can still switch out the implementations in unit tests with stubs if we want to.

Reply:
I meant you are no longer required to implement stubs for interface methods you are not interested in when using reduced private interfaces.